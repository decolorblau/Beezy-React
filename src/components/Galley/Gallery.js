import Card from "../Card/Card";
import { NavLink } from "react-router-dom";
import "./Gallery.scss";
import PropTypes from "prop-types";
import { useContext } from "react";
import MarvelContext from "../../store/contexts/MarvelContext";
import { css } from "@emotion/react";
import ClipLoader from "react-spinners/ClipLoader";

const override = css`
  display: block;
  margin: 0 auto;
  border-color: red;
`;

const Gallery = () => {
  const { comics, isLoading } = useContext(MarvelContext);

  return isLoading ? (
    <div className="gallery-empty-container">
      <ClipLoader loading={isLoading} css={override} size={120} />
    </div>
  ) : comics && comics.length > 0 ? (
    <>
      <div className="gallery-container" title="Comics">
        {comics.map((comic) => (
          <NavLink key={comic.id} to={`comic/${comic.id}`}>
            <Card key={comic.id} marvelComic={comic} />
          </NavLink>
        ))}
      </div>
    </>
  ) : (
    <div className="gallery-empty-container" title="No Comics">
      <p>No comics available</p>
    </div>
  );
};

Gallery.propTypes = {
  comics: PropTypes.array,
};

export default Gallery;
