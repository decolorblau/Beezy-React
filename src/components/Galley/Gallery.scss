.gallery-title {
  font-family: Roboto, sans-serif;
  font-size: 20px;
  color: #dcdcdc;
  letter-spacing: 3px;
  font-weight: normal;
  margin-top: 50px;
  margin-bottom: 10px;
}

.gallery-container {
  display: grid;
  grid-template-columns: repeat(3, 1fr);
  grid-template-rows: auto;
  padding: 0;
  max-width: 1450px;
}

@media (min-width: 768px) {
  .gallery-container {
    grid-template-columns: repeat(4, 1fr);
  }
  .gallery-title {
    font-size: 26px;
  }
}

@media (min-width: 1024px) {
  .gallery-container {
    grid-template-columns: repeat(6, 1fr);
  }
  .gallery-title {
    font-size: 35px;
  }
}
