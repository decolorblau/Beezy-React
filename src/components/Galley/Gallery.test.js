import Gallery from "./Gallery";
import { BrowserRouter as Router } from "react-router-dom";
import MarvelContext from "../../store/contexts/MarvelContext";
import { render, screen } from "@testing-library/react";
import ReactTestRenderer from "react-test-renderer";

describe("Given the component Gallery", () => {
  describe("When it receives an array of comics", () => {
    test("Then it should render div tag with gallery-container class", () => {
      const comics = {
        comics: [
          {
            id: 2345,
            title: "Avengers",
            format: "comic",
            thumbnail: {
              path: "https://image.tmdb.org/t/p/w500",
              extension: "jpg",
            },
            prices: [{ price: 3.99 }],
          },
          {
            id: 2375,
            title: "Avengers 2",
            format: "comic",
            thumbnail: {
              path: "https://image.tmdb.org/t/p/w500",
              extension: "jpg",
            },
            prices: [{ price: 2.99 }],
          },
        ],
      };
      const expected = "gallery-container";

      render(
        <MarvelContext.Provider value={comics}>
          <Router>
            <Gallery />
          </Router>
        </MarvelContext.Provider>
      );

      const galleryElement = screen.getByTitle("Comics");

      expect(galleryElement.className).toBe(expected);
    });
    test("Then it should render as many cards s as the array length", () => {
      const comics = {
        comics: [
          {
            id: 2345,
            title: "Avengers",
            format: "comic",
            thumbnail: {
              path: "https://image.tmdb.org/t/p/w500",
              extension: "jpg",
            },
            prices: [{ price: 3.99 }],
          },
          {
            id: 2375,
            title: "Avengers 2",
            format: "comic",
            thumbnail: {
              path: "https://image.tmdb.org/t/p/w500",
              extension: "jpg",
            },
            prices: [{ price: 2.99 }],
          },
        ],
      };

      render(
        <MarvelContext.Provider value={comics}>
          <Router>
            <Gallery />
          </Router>
        </MarvelContext.Provider>
      );

      const galleryElement = screen.getAllByRole("heading");

      expect(galleryElement).toHaveLength(comics.comics.length);
    });
    test("Then it should render gallery of comics", () => {
      const comics = {
        comics: [
          {
            id: 2345,
            title: "Avengers",
            format: "comic",
            thumbnail: {
              path: "https://image.tmdb.org/t/p/w500",
              extension: "jpg",
            },
            prices: [{ price: 3.99 }],
          },
          {
            id: 2375,
            title: "Avengers 2",
            format: "comic",
            thumbnail: {
              path: "https://image.tmdb.org/t/p/w500",
              extension: "jpg",
            },
            prices: [{ price: 2.99 }],
          },
        ],
      };

      const galleryComponent = ReactTestRenderer.create(
        <MarvelContext.Provider value={comics}>
          <Router>
            <Gallery />
          </Router>
        </MarvelContext.Provider>
      );

      expect(galleryComponent.toJSON()).toMatchSnapshot();
    });
  });
  describe("When it receives an empty array of comics", () => {
    test("It should render 'No comics available'", () => {
      const comics = {
        comics: [],
      };
      const expected = "gallery-empty-container";

      render(
        <MarvelContext.Provider value={comics}>
          <Router>
            <Gallery />
          </Router>
        </MarvelContext.Provider>
      );

      const emptyGalleryElement = screen.getByTitle("No Comics");

      expect(emptyGalleryElement.className).toBe(expected);
    });
  });
});
