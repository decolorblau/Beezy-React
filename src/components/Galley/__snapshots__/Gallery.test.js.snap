// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given the component Gallery When it receives an array of comics Then it should render gallery of comics 1`] = `
<div
  className="gallery-container"
  title="Comics"
>
  <a
    className=""
    href="/2345"
    onClick={[Function]}
  >
    <section
      className="card"
    >
      <img
        alt="Poster of Avengers"
        className="card__image"
        src="https://image.tmdb.org/t/p/w500.jpg"
      />
      <div
        className="card__info"
      >
        <h3
          className="card__title"
        >
          Avengers
        </h3>
        <div
          className="card__text"
        >
          <p
            className="card__format"
          >
            comic
          </p>
          <p
            className="card__price"
          >
            Price: 
            3.99
             €
          </p>
        </div>
      </div>
    </section>
  </a>
  <a
    className=""
    href="/2375"
    onClick={[Function]}
  >
    <section
      className="card"
    >
      <img
        alt="Poster of Avengers 2"
        className="card__image"
        src="https://image.tmdb.org/t/p/w500.jpg"
      />
      <div
        className="card__info"
      >
        <h3
          className="card__title"
        >
          Avengers 2
        </h3>
        <div
          className="card__text"
        >
          <p
            className="card__format"
          >
            comic
          </p>
          <p
            className="card__price"
          >
            Price: 
            2.99
             €
          </p>
        </div>
      </div>
    </section>
  </a>
</div>
`;
